### PR TITLE
EPMRPP-37506 || Integrations refactoring api

### PR DESCRIPTION
**Pull request by Yumfriez**

### CODE REVIEW

File: src/main/java/com/epam/ta/reportportal/auth/permissions/Permissions.java [hunk 1]
```diff
@@ -33,20 +33,22 @@ private Permissions() {
 
 	public static final String ALLOWED_TO_REPORT = "hasPermission(#projectName.toLowerCase(), 'reporterPermission')";
 
-	public static final String PROJECT_MANAGER = "hasPermission(#projectName.toLowerCase(), 'projectManagerPermission')";
+	public static final String ASSIGNED_TO_PROJECT = "hasPermission(#projectName.toLowerCase(), 'isAssignedToProject')";
+
+	public static final String PROJECT_MANAGER =
+			"hasPermission(#projectName.toLowerCase(), 'projectManagerPermission')" + " && " + ASSIGNED_TO_PROJECT;
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/auth/permissions/Permissions.java]: maybe the condition that project manager is assigned to project should be checked inside 'projectManagerPermission' evaluator?

File: src/main/java/com/epam/ta/reportportal/core/admin/ServerAdminHandlerImpl.java [hunk 2]
```diff
@@ -77,87 +66,17 @@ public ServerAdminHandlerImpl(BasicTextEncryptor simpleEncryptor, ServerSettings
 
 	@Override
 	public ServerSettingsResource getServerSettings() {
-		return ServerSettingsConverter.TO_RESOURCE.apply(serverSettingsRepository.findAll());
+		return null;
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/admin/ServerAdminHandlerImpl.java]: wtf?

File: src/main/java/com/epam/ta/reportportal/core/integration/CreateIntegrationHandler.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import com.epam.ta.reportportal.auth.ReportPortalUser;
+import com.epam.ta.reportportal.ws.model.OperationCompletionRS;
+import com.epam.ta.reportportal.ws.model.integration.UpdateIntegrationRQ;
+
+/**
+ * @author <a href="mailto:ivan_budayeu@epam.com">Ivan Budayeu</a>
+ */
+public interface CreateIntegrationHandler {
+
+	OperationCompletionRS createGlobalIntegration(UpdateIntegrationRQ updateRequest);
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/integration/CreateIntegrationHandler.java]: let's replace global with server

File: src/main/java/com/epam/ta/reportportal/core/integration/DeleteIntegrationHandler.java [hunk 4]
```diff
@@ -24,17 +24,29 @@
  */
 public interface DeleteIntegrationHandler {
 
+	/**
+	 * @param integrationId Integration id
+	 * @return Response data
+	 */
+	OperationCompletionRS deleteGlobalIntegration(Long integrationId);
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/integration/DeleteIntegrationHandler.java]: global -> server

File: src/main/java/com/epam/ta/reportportal/core/integration/GetIntegrationHandler.java [hunk 5]
```diff
@@ -29,5 +29,7 @@
 	 * @param projectDetails {@link com.epam.ta.reportportal.auth.ReportPortalUser.ProjectDetails}
 	 * @return {@link IntegrationResource}
 	 */
-	IntegrationResource getIntegrationById(Long integrationId, ReportPortalUser.ProjectDetails projectDetails);
+	IntegrationResource getProjectIntegrationById(Long integrationId, ReportPortalUser.ProjectDetails projectDetails);
+	
+	IntegrationResource getGlobalIntegrationById(Long integrationId);
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/integration/GetIntegrationHandler.java]: global -> server

File: src/main/java/com/epam/ta/reportportal/core/configs/IntegrationConfig.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+	private ApplicationContext applicationContext;
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = applicationContext;
+	}
+
+	@Bean
+	public Map<ReportPortalIntegrationEnum, IntegrationService> integrationServiceMapping() {
+		return ImmutableMap.<ReportPortalIntegrationEnum, IntegrationService>builder().put(ReportPortalIntegrationEnum.JIRA,
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/configs/IntegrationConfig.java]: what about rally?

File: src/main/java/com/epam/ta/reportportal/core/integration/impl/CreateIntegrationHandlerImpl.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
+package com.epam.ta.reportportal.core.integration.impl;
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/integration/impl/CreateIntegrationHandlerImpl.java]: headers missed everywhere

File: src/main/java/com/epam/ta/reportportal/core/integration/impl/CreateIntegrationHandlerImpl.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+		integrationRepository.save(integration);
+
+		IntegrationType integrationType = integrationTypeRepository.findByIntegrationGroup(integration.getType().getIntegrationGroup())
+				.orElseThrow(() -> new ReportPortalException(ErrorType.INTEGRATION_NOT_FOUND,
+						"Unknown integration group - " + integration.getType().getIntegrationGroup()
+				));
+
+		integrationRepository.updateIntegrationGroupEnabledState(updateRequest.getEnabled(), integrationType.getId());
+
+		return new OperationCompletionRS("Email integration with id = " + integration.getId() + " has been successfully created.");
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/integration/impl/CreateIntegrationHandlerImpl.java]: Email integration?

File: src/main/java/com/epam/ta/reportportal/core/integration/impl/CreateIntegrationHandlerImpl.java [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+		Integration integration = integrationServiceMapping.get(reportPortalIntegration)
+				.createProjectIntegration(updateRequest.getIntegrationName(), projectDetails, updateRequest.getIntegrationParams());
+
+		integration.setEnabled(updateRequest.getEnabled());
+
+		integration.setProject(project);
+
+		integrationRepository.save(integration);
+
+		return new OperationCompletionRS("Email integration with id = " + integration.getId() + " has been successfully created.");
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/integration/impl/CreateIntegrationHandlerImpl.java]: email

File: src/main/java/com/epam/ta/reportportal/core/integration/impl/DeleteIntegrationHandlerImpl.java [hunk 10]
```diff
@@ -59,7 +80,7 @@ public OperationCompletionRS deleteIntegration(Long integrationId, ReportPortalU
 		integrationRepository.delete(integration);
 
 		messageBus.publishActivity(new IntegrationDeletedEvent(TO_ACTIVITY_RESOURCE.apply(integration), user.getUserId()));
-		return new OperationCompletionRS("ExternalSystem with ID = '" + integrationId + "' is successfully deleted.");
+		return new OperationCompletionRS("Integration with ID = '" + integrationId + "' have been successfully deleted.");
```
Comment (pbortnik) [inline src/main/java/com/epam/ta/reportportal/core/integration/impl/DeleteIntegrationHandlerImpl.java]: has been?

File: src/main/java/com/epam/ta/reportportal/auth/permissions/Permissions.java (hunk 1 above)
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/auth/permissions/Permissions.java]: my bad, that condition is checked in the 'BaseProjectPermission'

File: src/main/java/com/epam/ta/reportportal/ws/controller/IntegrationController.java [hunk 11]
```diff
@@ -18,175 +18,129 @@
...
+		return deleteIntegrationHandler.deleteProjectIntegration(integrationId,
 				extractProjectDetails(user, EntityUtils.normalizeId(projectName)),
 				user
 		);
 	}
 
-	@Transactional(readOnly = true)
-	@GetMapping(value = "/{integrationId}/ticket/{ticketId}")
+	@Transactional
+	@DeleteMapping(value = "/{projectName}/clear")
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/ws/controller/IntegrationController.java]: renamed

File: src/main/java/com/epam/ta/reportportal/util/email/MailServiceFactory.java [hunk 12]
```diff
@@ -67,7 +69,7 @@ public MailServiceFactory(TemplateEngine templateEngine, BasicTextEncryptor encr
 	 * @return Built email service
 	 */
 	private Optional<EmailService> getEmailService(Integration emailIntegration, List<ServerSettings> serverSettings) {
-		return getEmailService(serverSettings).flatMap(service -> {
+		return getEmailService(null).flatMap(service -> {
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/util/email/MailServiceFactory.java]: done

File: src/main/java/com/epam/ta/reportportal/util/email/MailServiceFactory.java [hunk 13]
```diff
@@ -125,7 +133,7 @@ public MailServiceFactory(TemplateEngine templateEngine, BasicTextEncryptor encr
 	 * @return Built email service
 	 */
 	public Optional<EmailService> getDefaultEmailService() {
-		return getEmailService(settingsRepository.findAll());
+		return getEmailService(null);
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/util/email/MailServiceFactory.java]: done

File: src/main/java/com/epam/ta/reportportal/util/email/MailServiceFactory.java [hunk 14]
```diff
@@ -162,7 +169,7 @@ public EmailService getDefaultEmailService(Integration integration, boolean chec
 	 */
 	@Transactional(propagation = Propagation.REQUIRES_NEW)
 	public EmailService getDefaultEmailService(boolean checkConnection) {
-		EmailService emailService = getEmailService(settingsRepository.findAll()).orElseThrow(() -> emailConfigurationFail(null));
+		EmailService emailService = getEmailService(null).orElseThrow(() -> emailConfigurationFail(null));
```
[the remark this answers is not in the document]
Comment (Yumfriez) [reply src/main/java/com/epam/ta/reportportal/util/email/MailServiceFactory.java]: done

**Pull request merged**